File: mobile/lib/models/fantasy.dart
// نماذج «فريقي» — لاعبٌ في السوق، وتشكيلة، وصفٌّ في نقاط الجولة.
//
// «اللاعب» هنا كرويٌّ في الملعب لا مستخدمٌ عندنا (ذاك Player في
// models/player.dart). الاسمان متشابهان والفرق جوهري، فاللاحقة
// Fantasy تحسم الالتباس عند القراءة بلا أن تُقرأ الاستيرادات.
import '../format.dart';

/// المراكز كما يخزّنها الخادم — مفاتيح قواعد لا نصوص عرض.
enum FantasyPosition { goalkeeper, defender, midfielder, attacker }

extension FantasyPositionX on FantasyPosition {
  static FantasyPosition parse(String? raw) => switch (raw) {
        'Goalkeeper' => FantasyPosition.goalkeeper,
        'Defender' => FantasyPosition.defender,
        'Midfielder' => FantasyPosition.midfielder,
        _ => FantasyPosition.attacker,
      };

  String get wire => switch (this) {
        FantasyPosition.goalkeeper => 'Goalkeeper',
        FantasyPosition.defender => 'Defender',
        FantasyPosition.midfielder => 'Midfielder',
        FantasyPosition.attacker => 'Attacker',
      };

  /// الاسم العربي الكامل — لعناوين الأقسام ورسائل الخطأ.
  String get label => switch (this) {
        FantasyPosition.goalkeeper => 'حراسة',
        FantasyPosition.defender => 'دفاع',
        FantasyPosition.midfielder => 'وسط',
        FantasyPosition.attacker => 'هجوم',
      };

  /// حرفٌ واحد لشارة البطاقة على الملعب — المساحة هناك ضيّقة.
  String get short => switch (this) {
        FantasyPosition.goalkeeper => 'حا',
        FantasyPosition.defender => 'مد',
        FantasyPosition.midfielder => 'وس',
        FantasyPosition.attacker => 'مه',
      };
}

class FantasyPlayer {
  final int id;
  final String name;
  final String? photoUrl;
  final FantasyPosition position;
  final int? teamId;
  final String teamName;
  final String? teamLogo;
  final double price;
  final int totalPoints;

  // حالته داخل تشكيلتي — لاغية لمن هو في السوق وحده.
  final bool onBench;
  final bool isCaptain;
  final bool isVice;

  const FantasyPlayer({
    required this.id,
    required this.name,
    required this.position,
    required this.price,
    this.photoUrl,
    this.teamId,
    this.teamName = '',
    this.teamLogo,
    this.totalPoints = 0,
    this.onBench = false,
    this.isCaptain = false,
    this.isVice = false,
  });

  factory FantasyPlayer.fromJson(Map<String, dynamic> j) => FantasyPlayer(
        id: j['player_id'] as int,
        name: (j['name'] as String?) ?? 'لاعب',
        photoUrl: j['photo_url'] as String?,
        position: FantasyPositionX.parse(j['position'] as String?),
        teamId: j['team_id'] as int?,
        teamName: (j['team_name'] as String?) ?? '',
        teamLogo: j['team_logo'] as String?,
        // السعر يصل نصّاً من NUMERIC — عمود مالي لا يُرسل عائماً.
        price: double.tryParse('${j['price']}') ?? 0,
        totalPoints: (j['total_points'] as int?) ?? 0,
        onBench: j['on_bench'] == true,
        isCaptain: j['is_captain'] == true,
        isVice: j['is_vice'] == true,
      );

  FantasyPlayer copyWith({bool? onBench, bool? isCaptain, bool? isVice}) =>
      FantasyPlayer(
        id: id,
        name: name,
        photoUrl: photoUrl,
        position: position,
        teamId: teamId,
        teamName: teamName,
        teamLogo: teamLogo,
        price: price,
        totalPoints: totalPoints,
        onBench: onBench ?? this.onBench,
        isCaptain: isCaptain ?? this.isCaptain,
        isVice: isVice ?? this.isVice,
      );

  /// السعر كما يُعرض: خانة عشرية واحدة بأرقام غربية دائماً.
  String get priceLabel => Fmt.number(price, decimals: 1);
}

class FantasySquad {
  final String id;
  final String? name;
  final int? clubTeamId;
  final String formation;
  final double budgetLeft;
  final int freeTransfers;
  final int totalPoints;
  final List<FantasyPlayer> players;

  const FantasySquad({
    required this.id,
    required this.formation,
    required this.budgetLeft,
    required this.players,
    this.name,
    this.clubTeamId,
    this.freeTransfers = 1,
    this.totalPoints = 0,
  });

  factory FantasySquad.fromJson(Map<String, dynamic> j) => FantasySquad(
        id: '${j['id']}',
        name: j['name'] as String?,
        clubTeamId: j['club_team_id'] as int?,
        formation: (j['formation'] as String?) ?? '4-4-2',
        budgetLeft: double.tryParse('${j['budget_left']}') ?? 0,
        freeTransfers: (j['free_transfers'] as int?) ?? 1,
        totalPoints: (j['total_points'] as int?) ?? 0,
        players: ((j['players'] as List?) ?? const [])
            .map((p) => FantasyPlayer.fromJson(p as Map<String, dynamic>))
            .toList(),
      );

  List<FantasyPlayer> get starters =>
      players.where((p) => !p.onBench).toList(growable: false);
  List<FantasyPlayer> get bench =>
      players.where((p) => p.onBench).toList(growable: false);
}

/// حدود اللعبة كما يقولها الخادم — لا تُكتب أرقامها هنا.
///
/// التشكيلة خمسة عشر والميزانية مئة اليوم، وقد تتغيّر من اللوحة
/// غداً. نسخةٌ ثانية من الأرقام في التطبيق تعني شاشةً تمنع ما
/// يقبله الخادم، أو تعد بما يرفضه.
class FantasyRules {
  final int squadSize;
  final int starters;
  final double budget;
  final int minFromClub;
  final int maxFromClub;
  final List<String> formations;
  final Map<FantasyPosition, int> quota;

  const FantasyRules({
    required this.squadSize,
    required this.starters,
    required this.budget,
    required this.minFromClub,
    required this.maxFromClub,
    required this.formations,
    required this.quota,
  });

  static const fallback = FantasyRules(
    squadSize: 15,
    starters: 11,
    budget: 100,
    minFromClub: 3,
    maxFromClub: 3,
    formations: ['4-4-2', '4-3-3', '4-5-1', '3-5-2', '3-4-3', '5-3-2', '5-4-1'],
    quota: {
      FantasyPosition.goalkeeper: 2,
      FantasyPosition.defender: 5,
      FantasyPosition.midfielder: 5,
      FantasyPosition.attacker: 3,
    },
  );

  factory FantasyRules.fromJson(Map<String, dynamic>? rules,
      List? formations, Map<String, dynamic>? quota) {
    if (rules == null) return fallback;
    return FantasyRules(
      squadSize: (rules['squad_size'] as int?) ?? 15,
      starters: (rules['starters'] as int?) ?? 11,
      budget: double.tryParse('${rules['budget']}') ?? 100,
      minFromClub: (rules['min_from_club'] as int?) ?? 3,
      maxFromClub: (rules['max_from_club'] as int?) ?? 3,
      formations: formations?.cast<String>() ?? fallback.formations,
      quota: quota == null
          ? fallback.quota
          : {
              for (final e in quota.entries)
                FantasyPositionX.parse(e.key): e.value as int,
            },
    );
  }

  /// عدد كل خط في الأساسي حسب الخطة — «4-4-2» ← ٤ دفاع، ٤ وسط، ٢ هجوم.
  static Map<FantasyPosition, int> shapeOf(String formation) {
    final parts = formation.split('-').map(int.tryParse).toList();
    if (parts.length != 3 || parts.any((p) => p == null)) {
      return shapeOf('4-4-2');
    }
    return {
      FantasyPosition.goalkeeper: 1,
      FantasyPosition.defender: parts[0]!,
      FantasyPosition.midfielder: parts[1]!,
      FantasyPosition.attacker: parts[2]!,
    };
  }
}

/// صفٌّ في شاشة نقاط الجولة.
class FantasyRoundRow {
  final FantasyPlayer player;
  final int points;
  final int multiplier;
  final bool autoSubbed;
  final bool onBench;

  const FantasyRoundRow({
    required this.player,
    required this.points,
    required this.multiplier,
    required this.autoSubbed,
    required this.onBench,
  });

  factory FantasyRoundRow.fromJson(Map<String, dynamic> j) => FantasyRoundRow(
        player: FantasyPlayer.fromJson(j),
        points: (j['points'] as int?) ?? 0,
        multiplier: (j['multiplier'] as int?) ?? 1,
        autoSubbed: j['auto_subbed'] == true,
        onBench: j['on_bench'] == true,
      );
}

/// صفٌّ في عرش «فريقي».
class FantasyRankRow {
  final int rank;
  final String name;
  final String? avatarUrl;
  final String? clubName;
  final String? clubLogo;
  final int totalPoints;
  final bool isMe;

  const FantasyRankRow({
    required this.rank,
    required this.name,
    required this.totalPoints,
    required this.isMe,
    this.avatarUrl,
    this.clubName,
    this.clubLogo,
  });

  factory FantasyRankRow.fromJson(Map<String, dynamic> j) => FantasyRankRow(
        rank: (j['rank'] as int?) ?? 0,
        name: (j['name'] as String?) ?? 'مشجع',
        avatarUrl: j['avatar_url'] as String?,
        clubName: j['club_name'] as String?,
        clubLogo: j['club_logo'] as String?,
        totalPoints: (j['total_points'] as int?) ?? 0,
        isMe: j['is_me'] == true,
      );
}

/// حمولة شاشة السوق كاملة: الحدود واللاعبون في ردٍّ واحد.
///
/// طلبٌ واحد لا اثنان: الشاشة لا تُرسم قبل وصول الاثنين معاً،
/// وطلبان متتاليان يعنيان دوّارةً مرتين على شبكةٍ بطيئة.
class FantasyMarket {
  final FantasyRules rules;
  final List<FantasyPlayer> players;
  final int? leagueId;
  final String? leagueName;
  final bool followRequired;

  const FantasyMarket({
    required this.rules,
    required this.players,
    this.leagueId,
    this.leagueName,
    this.followRequired = false,
  });

  factory FantasyMarket.fromJson(Map<String, dynamic> j) => FantasyMarket(
        rules: FantasyRules.fromJson(
          j['rules'] as Map<String, dynamic>?,
          j['formations'] as List?,
          j['quota'] as Map<String, dynamic>?,
        ),
        players: ((j['players'] as List?) ?? const [])
            .map((p) => FantasyPlayer.fromJson(p as Map<String, dynamic>))
            .toList(),
        leagueId: (j['league'] as Map<String, dynamic>?)?['id'] as int?,
        leagueName: (j['league'] as Map<String, dynamic>?)?['name'] as String?,
        followRequired: j['follow_required'] == true,
      );
}

File: server/src/routes/fantasy.js
// routes/fantasy — «فريقي»: السوق والتشكيلة والجولة والعرش.
//
// كله خلف requireAuth: التشكيلة ملكُ صاحبها، والسوق نفسه يعرض
// أسعاراً تخصّ لعبةً لا يدخلها ضيف. والعرش وحده كان يمكن أن يكون
// عاماً، لكنه يُقرأ من داخل التبويب فلا فائدة من فتحه.
//
// والقواعد كلها في fantasySquadService: هذا المسار يفهم HTTP
// ويحوّل أخطاء القواعد إلى 400 برسالتها العربية كما هي.
const express = require('express');
const requireAuth = require('../middleware/requireAuth');
const leagueRepo = require('../repositories/leagueRepo');
const championRepo = require('../repositories/championRepo');
const fixtureRepo = require('../repositories/fixtureRepo');
const fantasyRepo = require('../repositories/fantasyRepo');
const squadService = require('../services/fantasySquadService');
const logger = require('../utils/logger');

const router = express.Router();
router.use(requireAuth);

/**
 * الدوري المقصود: المطلوب إن كان يتابعه، وإلا أول متابعاته.
 *
 * نفس حارس routes/round: التشكيلة تُبنى من لاعبي دوريٍّ يتابعه،
 * ومن لا يتابع شيئاً لا تشكيلة له — والجواب دعوةٌ للمتابعة لا
 * قائمةٌ فارغة لا تشرح نفسها.
 */
async function resolveLeague(req) {
  const followed = await championRepo.followedIds(req.userId);
  const leagues = (await leagueRepo.findEnabled()).filter(
    (l) => l.in_app && followed.includes(l.id)
  );
  if (!leagues.length) return null;
  const asked = Number(req.query.league || req.body?.league);
  return leagues.find((l) => l.id === asked) || leagues[0];
}

const FOLLOW_REQUIRED = { follow_required: true, league: null, squad: null };

// GET /api/fantasy/market — سوق اللاعبين
router.get('/market', async (req, res) => {
  const league = await resolveLeague(req);
  if (!league) return res.json({ ...FOLLOW_REQUIRED, players: [] });

  const players = await fantasyRepo.market(league.id, league.season, {
    position: ['Goalkeeper', 'Defender', 'Midfielder', 'Attacker']
      .includes(req.query.position) ? req.query.position : null,
    teamId: /^\d+$/.test(req.query.team) ? Number(req.query.team) : null,
    search: String(req.query.q || '').trim() || null,
  });

  res.json({
    league: { id: league.id, name: league.name, logo_url: `/logos/league-${league.id}.png` },
    rules: squadService.RULES,
    formations: Object.keys(squadService.FORMATIONS),
    quota: squadService.SQUAD_QUOTA,
    players,
  });
});

// GET /api/fantasy/squad — تشكيلتي
router.get('/squad', async (req, res) => {
  const league = await resolveLeague(req);
  if (!league) return res.json(FOLLOW_REQUIRED);

  const squad = await fantasyRepo.findSquad(req.userId, league.id, league.season);
  res.json({
    league: { id: league.id, name: league.name },
    rules: squadService.RULES,
    formations: Object.keys(squadService.FORMATIONS),
    squad: squad && {
      id: squad.id,
      name: squad.name,
      club_team_id: squad.club_team_id,
      formation: squad.formation,
      budget_left: Number(squad.budget_left),
      // قيمة اللاعبين بأسعار اليوم. مع المحفظة تعطي «قيمة فريقك»،
      // وهي المقياس الثاني في اللعبة بعد النقاط: من ينمو فريقه
      // يشتري ما لا يستطيعه غيره.
      squad_value: Number(squad.squad_value),
      free_transfers: squad.free_transfers,
      wildcards_used: squad.wildcards_used,
      total_points: squad.total_points,
      players: await fantasyRepo.squadPlayers(squad.id),
    },
  });
});

// PUT /api/fantasy/squad — حفظ التشكيلة كاملة
//
// كاملة لا جزئية: قواعد اللعبة (الحصص، الخطة، الأندية، الميزانية)
// لا تُفحص إلا على خمسة عشر معاً، وتعديلٌ جزئي يعني حالةً وسيطة
// مكسورة يجب أن نقرّر ماذا نفعل بها — ولا جواب صحيح.
router.put('/squad', async (req, res) => {
  const league = await resolveLeague(req);
  if (!league) return res.status(400).json({ error: 'تابِع دورياً أولاً كي تبني تشكيلتك.' });

  const picks = Array.isArray(req.body?.players) ? req.body.players : [];
  const formation = String(req.body?.formation || '');
  const clubTeamId = Number(req.body?.club_team_id) || null;

  const ids = picks.map((p) => Number(p.player_id)).filter(Number.isInteger);
  const rows = await fantasyRepo.playersByIds(ids);
  const byId = new Map(rows.map((r) => [r.player_id, r]));

  const cleaned = picks.map((p) => ({
    player_id: Number(p.player_id),
    on_bench: !!p.on_bench,
    is_captain: !!p.is_captain,
    is_vice: !!p.is_vice,
  }));

  try {
    // المحفظة قبل التحقّق: قاعدتها حركةُ بيعٍ وشراء لا مجموع
    // أسعار اليوم، فلا يمكن للفحص أن يعرفها وحده.
    const plan = await fantasyRepo.planTransaction({
      userId: req.userId,
      leagueId: league.id,
      season: league.season,
      picks: cleaned,
      prices: byId,
      budget: squadService.RULES.budget,
    });

    const result = squadService.validateSquad(cleaned, byId, {
      formation,
      clubTeamId,
      leagueId: league.id,
      wallet: plan.wallet,
    });

    // ترتيب الدكّة من الخدمة لا من العميل: هي من تعرف أن الحارس
    // البديل يبقى آخراً، والعميل قد يرسل ترتيباً يكسر التبديل.
    const benchOrder = new Map(result.bench.map((b, i) => [b.player_id, i]));

    // الجولة التي تُسجَّل فيها الانتقالات: الجارية — وهي نافذة
    // السوق المفتوحة الآن.
    const rounds = await fixtureRepo.roundsFor(league.id, league.season);
    const openRound = (rounds.find((r) => r.open > 0) || rounds[rounds.length - 1])?.round;

    await fantasyRepo.saveSquad({
      userId: req.userId,
      leagueId: league.id,
      season: league.season,
      clubTeamId,
      formation,
      name: req.body?.name,
      wallet: plan.wallet,
      value: result.value,
      sold: plan.sold,
      bought: plan.bought.map((p) => ({
        player_id: p.player_id,
        price: byId.get(p.player_id)?.price ?? 0,
      })),
      round: openRound,
      rows: result.rows.map((r) => ({
        player_id: r.player_id,
        on_bench: r.on_bench,
        bench_order: benchOrder.get(r.player_id) ?? 0,
        is_captain: r.is_captain,
        is_vice: r.is_vice,
        price: r.price,
      })),
    });

    const squad = await fantasyRepo.findSquad(req.userId, league.id, league.season);
    res.json({
      squad: {
        id: squad.id,
        formation: squad.formation,
        club_team_id: squad.club_team_id,
        budget_left: Number(squad.budget_left),
        squad_value: Number(squad.squad_value),
        free_transfers: squad.free_transfers,
        total_points: squad.total_points,
        players: await fantasyRepo.squadPlayers(squad.id),
      },
    });
  } catch (err) {
    if (err.code === 'FANTASY_RULE') return res.status(400).json({ error: err.message });
    throw err;
  }
});

// GET /api/fantasy/round — نقاط الجولة، لاعباً لاعباً
//
// من الجولة المجمّدة لا من التشكيلة الحالية: من بدّل كابتنه بعد
// المباراة الأولى يجب أن يرى الجولة كما لُعبت لا كما صارت.
router.get('/round', async (req, res) => {
  const league = await resolveLeague(req);
  if (!league) return res.json({ ...FOLLOW_REQUIRED, round: null, players: [] });

  const squad = await fantasyRepo.findSquad(req.userId, league.id, league.season);
  if (!squad) return res.json({ league: { id: league.id }, squad: null, round: null, players: [] });

  const rounds = await fixtureRepo.roundsFor(league.id, league.season);
  const asked = String(req.query.round || '');
  const round = rounds.some((r) => r.round === asked)
    ? asked
    : (rounds.find((r) => r.open > 0) || rounds[rounds.length - 1])?.round || null;

  const lineup = round ? await fantasyRepo.roundLineup(squad.id, round) : [];
  const names = await fantasyRepo.squadPlayers(squad.id);
  const byId = new Map(names.map((n) => [n.player_id, n]));

  res.json({
    league: { id: league.id, name: league.name },
    round,
    rounds: rounds.map((r) => r.round),
    total: lineup.reduce((sum, l) => sum + l.points, 0),
    players: lineup.map((l) => ({ ...l, ...(byId.get(l.player_id) || {}) })),
  });
});

// GET /api/fantasy/leaderboard — عرش «فريقي»
router.get('/leaderboard', async (req, res) => {
  const league = await resolveLeague(req);
  if (!league) return res.json({ ...FOLLOW_REQUIRED, entries: [] });

  const entries = await fantasyRepo.leaderboard(league.id, league.season);
  res.json({
    league: { id: league.id, name: league.name },
    me: entries.find((e) => e.user_id === req.userId) || null,
    entries: entries.map((e) => ({
      rank: Number(e.rank),
      name: e.name || e.display_name || 'مشجع',
      avatar_url: e.avatar_url,
      club_name: e.club_name,
      club_logo: e.club_logo,
      total_points: e.total_points,
      is_me: e.user_id === req.userId,
    })),
  });
});

module.exports = router;
